src/i_pcsound.c

```diff
@@ -0,0 +1,475 @@
+// Heretic source code, where there are remnants of this left over
+// from Doom.
+
+static boolean IsDisabledSound(sfxinfo_t *sfxinfo)
+{
+    int i;
+    const char *disabled_sounds[] = {
+        "posact",
+        "bgact",
+        "dmact",
+        "dmpain",
+        "popain",
+        "sawidl",
+    };
+
+    for (i = 0; i < arrlen(disabled_sounds); ++i)
+    {
+        if (!strcmp(sfxinfo->name, disabled_sounds[i]))
+        {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+// Mixer function that does the PC speaker emulation
+
+static ALsizei AL_APIENTRY BufferCallback(void *userptr, void *data, ALsizei size)
+{
+    int16_t *leftptr;
+    int16_t *rightptr;
+    int16_t this_value;
+    int frequency;
+    int i;
+    int nsamples;
+
+    // Number of samples is quadrupled, because of 16-bit and stereo
+
+    nsamples = size / 4;
+
+    leftptr = (int16_t *) data;
+    rightptr = ((int16_t *) data) + 1;
+
+    // Fill the output buffer
+
+    for (i = 0; i < nsamples; ++i)
+    {
+        // Has this sound expired? If so, invoke the callback to get the next
+        // frequency.
+
+        while (current_remaining == 0) 
+        {
+            // Get the next frequency to play
+
+            GetFreq(&current_remaining, &frequency);
+
+            if (current_freq != frequency)
+            {
+                current_freq = frequency;
+                phase_offset = 0;
+            }
+
+            current_remaining = (current_remaining * mixing_freq) / 1000;
+        }
+
+        // Set the value for this sample.
+
+        if (current_freq == 0)
+        {
+            // Silence
+
+            this_value = 0;
+        }
+        else
+        {
+            int frac;
+
+            // Determine whether we are at a peak or trough in the current
+            // sound.  Multiply by 2 so that frac % 2 will give 0 or 1 
+            // depending on whether we are at a peak or trough.
+
+            frac = (phase_offset * current_freq * 2) / mixing_freq;
+
+            if ((frac % 2) == 0) 
+            {
+                this_value = SQUARE_WAVE_AMP;
+            }
+            else
+            {
+                this_value = -SQUARE_WAVE_AMP;
+            }
+
+            ++phase_offset;
+        }
+
+        --current_remaining;
+
+        // Use the same value for the left and right channels.
+
+        *leftptr = this_value;
+        *rightptr = this_value;
+
+        leftptr += 2;
+        rightptr += 2;
+    }
+
+    return size;
+}
+
+static void RegisterCallback(void)
+{
+    if (!alIsExtensionPresent("AL_SOFT_callback_buffer"))
+    {
+        I_Printf(VB_ERROR, "RegisterCallback: AL_SOFT_callback_buffer not found.");
+        return;
+    }
+    alBufferCallbackSOFT = FUNCTION_CAST(LPALBUFFERCALLBACKSOFT,
+                                         alGetProcAddress("alBufferCallbackSOFT"));
+
+    alGenBuffers(1, &callback_buffer);
+    alGenSources(1, &callback_source);
```

Comment:
So the PC speaker module handles it's own buffer (just one sound at a time?) and doesn't need to worry about the channel priority system or `snd_channels`, correct?

Reply:
It's only one sound at a time, but maybe we should worry about the priority system. Most of the time it works like vanilla, but for example, try [PL2.wad](https://www.doomworld.com/idgames/levels/doom2/megawads/pl2) `-playdemo demo2` in Woof and Crispy - there's a lot of difference in the PC Speaker sound there.

Reply:
I believe `I_PCS_AdjustSoundParams()` will need to have at least the vanilla distance check as shown here: https://github.com/fabiangreffrath/woof/blob/aabca29fcc5020d906a81add342c08fa04e29216/src/i_mbfsound.c#L69

Reply:
Does pl2.wad demo2 happen to start on map08?

Reply:
> Does pl2.wad demo2 happen to start on map08?

No, map15. demo1 is also sounds different.

So, I tried almost entire `I_MBF_AdjustSoundParams`, sounds much better:
```diff
diff --git a/src/i_pcsound.c b/src/i_pcsound.c
index f56c367..bc2f8a8 100644
--- a/src/i_pcsound.c
+++ b/src/i_pcsound.c
@@ -386,8 +386,59 @@ static boolean I_PCS_CacheSound(sfxinfo_t *sfx)
 static boolean I_PCS_AdjustSoundParams(const mobj_t *listener, const mobj_t *source,
                                       int chanvol, int *vol, int *sep, int *pri)
 {
+    fixed_t adx, ady, dist;
+
+    // haleyjd 05/29/06: allow per-channel volume scaling
+    *vol = (snd_SfxVolume * chanvol) / 15;
+
+    if (*vol < 1)
+        return false;
+    else if (*vol > 127)
+        *vol = 127;
+
+    *sep = NORM_SEP;
+
+    if (!source || source == players[displayplayer].mo)
+        return true;
+
+    // haleyjd 08/12/04: we cannot adjust a sound for a NULL listener.
+    if (!listener)
+        return true;
+
+    // calculate the distance to sound origin
+    //  and clip it if necessary
+    //
+    // killough 11/98: scale coordinates down before calculations start
+    // killough 12/98: use exact distance formula instead of approximation
+
+    adx = abs((listener->x >> FRACBITS) - (source->x >> FRACBITS));
+    ady = abs((listener->y >> FRACBITS) - (source->y >> FRACBITS));
+
+    if (ady > adx)
+        dist = adx, adx = ady, ady = dist;
+
+    dist = adx ? FixedDiv(adx, finesine[(tantoangle[FixedDiv(ady, adx) >> DBITS]
+                                         + ANG90) >> ANGLETOFINESHIFT]) : 0;
+
+    if (!dist)  // killough 11/98: handle zero-distance as special case
+        return true;
+
+    if (dist > S_CLIPPING_DIST >> FRACBITS)
+        return false;
+
+    // volume calculation
+    if (dist > S_CLOSE_DIST >> FRACBITS)
+        *vol = *vol * ((S_CLIPPING_DIST >> FRACBITS) - dist) / S_ATTENUATOR;
+
+    // haleyjd 09/27/06: decrease priority with volume attenuation
+    *pri = *pri + (127 - *vol);
+
+    if (*pri > 255) // cap to 255
+        *pri = 255;
+
     alSourcef(callback_source, AL_GAIN, (float)snd_SfxVolume / 15);
-    return true;
+
+    return (*vol > 0);
 }
 
 static void I_PCS_UpdateSoundParams(int channel, int volume, int separation)
```

Reply:
Done in https://github.com/fabiangreffrath/woof/pull/1200/commits/e08ef48093f6fb331dfd6fe7a07043b836ce7755

Reply:
Okay that's closer. Two things I noticed:

1. Something is still different in how sounds are able to overlap. Listen to the second beep in each example: [recordings.zip](https://github.com/fabiangreffrath/woof/files/12578608/recordings.zip)
2. Crispy starts the cyberdemon sound after the wipe finishes, Woof plays it immediately. This happens with regular sound (non-PC speaker) as well. I checked 10.5.1 and it's the same, so Woof has been like this for a long time. Maybe this is a topic for a different PR. I'm not sure which is correct.

Reply:
> Something is still different in how sounds are able to overlap. Listen to the second beep in each example: [recordings.zip](https://github.com/fabiangreffrath/woof/files/12578608/recordings.zip)

Interesting, I think Woof sounds obviously wrong. How do you get that beep sound? I have no idea where the bug might be.

> Crispy starts the cyberdemon sound after the wipe finishes, Woof plays it immediately...

There are a lot of differences between MBF and vanilla sound. At one time I completely copied the Crispy/Choco sound system, but we decided to keep the MBF system.

Reply:
> How do you get that beep sound?

When the player jumps over the stairs in pl2 demo2.

> There are a lot of differences between MBF and vanilla sound

Yeah, I don't think it's a problem either. For example, the different priority system MBF introduced completely alters the sound.